### PR TITLE
random uniform -> normal to initiate lobpcg and arpack in svds

### LINKED ISSUE #15152: BUG: initial approximations to eigenvectors in svds should not be with all positive components
### Describe your issue.

https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/_eigen/_svds.py at least in one place sets initial approximations to eigenvectors with all components random uniform on [0, 1]. This is a poor choice as common sense suggests that vectors with only positive components cover the positive quadrant only, not the whole vector space, are generally not so good to approximate eigenvectors. E.g., all example in the original MATLAB/Octave code, written for the paper, https://github.com/lobpcg/blopex/blob/master/blopex_tools/matlab/lobpcg/lobpcg.m use randn, e.g., line 97.
Poor initial approximations may make the solver running longer or even fail. 

The simplest solution is replacing all calls of uniform random with normal random. 

### Reproducing Code Example

```python
N/A
```

### Error message

```shell
N/A
```

### SciPy/NumPy/Python version information

N/A

### PR DESCRIPTION
**Pull request by lobpcg**
#### Reference issue
Closes gh-15152

#### What does this implement/fix?
https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/_eigen/_svds.py at least in one place sets initial approximations to eigenvectors with all components random uniform on [0, 1]. This is a poor choice as common sense suggests that vectors with only positive components cover the positive quadrant only, not the whole vector space, are generally not so good to approximate eigenvectors. E.g., all example in the original MATLAB/Octave code, written for the paper, https://github.com/lobpcg/blopex/blob/master/blopex_tools/matlab/lobpcg/lobpcg.m use randn, e.g., line 97.
Poor initial approximations may make the solver running longer or even fail.

The same arguments hold for ARPACK too

The simplest solution is replacing all calls of uniform random with normal random.

### CODE REVIEW

**Comment by lobpcg** [comment]
@ilayn @mdhaber @tupui @tylerjereddy @ogauthe @pv @evgueni-ovtchinnikov could someone please review this simple PR?

File: scipy/sparse/linalg/_eigen/_svds.py
Comment (tupui) [inline scipy/sparse/linalg/_eigen/_svds.py:291]: `standart_normal` should be used as `randn` is not advised in new code.

Also since this would break backward compatibility, we should use our RNG helper function from `_utils`. Fixing the seed to an arbitrary number like this is an abomination we should never have had 😅

**Review by tupui — CHANGES_REQUESTED** [review]
Thank you @lobpcg. Assuming breaking backward compatibility is fine here (and I don't know because it's not my area), I have some suggestions.

Note on our workflow: Please don't ping everyone at random. Waiting a few days to get a PR reviewed is normal as most maintainers are volunteers. Only after a while (a week or so) you should consider pinging maintainers, not a whole list but dedicated people you believe would have the required expertise (in this case not me for instance). Also we are in a release stage at the moment and any extra ping is creating unnecessary tension during the process.

**Comment by ogauthe** [comment]
I think backward compatibility is not really an issue here: if getting reproducible results matters one should anyway fix `v0` and then `random_state` is not use.

I agree with @tupui that we should let `check_random_state` do the job instead of hard coding the seed here.

Concerning the rationale, I agree positive coordinate is bad, but why not "just" shift those to get uniform coordinates between -0.5 and 0.5? Are there reasons to expect a normal distribution to give better results?

**Comment by rkern** [comment]
`standard_normal()` will be isotropic while `uniform(-x, x)` will have extra mass concentrated at the corners of the hyperbox. That said, ARPACK seems to use `uniform(-1, 1)`: 

https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_eigen/arpack/ARPACK/SRC/dgetv0.f#L225-L235

Just a quick note: `_iv()` does the `check_random_state()` coercion. So you can just use `random_state` after that call and remove the `rs_was_None` dance. I agree that the `random_state=None` behavior doesn't need to be kept.

**Comment by lobpcg** [comment]
@rkern @ogauthe @tupui thanks for your super-prompt relies! 

My goal was limited to a 2-liner quickie just changing all positive components into something more reasonable, hoping to get it merged fast and try to fix the other, more troubling, issue in this svds code - the use of tol *2 tolerance. 

I am unfamiliar with random seeding tools and not really up to the task of changing the logic of the code in this respect. With this in mind, could one of you please make specific editing suggestions that I could simply accept or better yet become a co-contributor and edit directly in this PR, aiming to have it finished fast?

File: scipy/sparse/linalg/_eigen/_svds.py
Comment (rkern) [inline scipy/sparse/linalg/_eigen/_svds.py]: This line can go, too.

Comment (rkern) [reply scipy/sparse/linalg/_eigen/_svds.py]: In a different PR, maybe.

**Comment by lobpcg** [comment]
@tupui - thanks for taking the lead! 
I am unsure why one instance has changed behavior and now fails, https://github.com/scipy/scipy/runs/6673308578?check_suite_focus=true claiming that a set of singular vectors is not orthonormal as far as I can see. But it appears that this is a complex arithmetic case for the solver ARPACK which to my recollection does not support complex? If so, it is more surprising why the test passes in other instances....

**Comment by ogauthe** [comment]
`svds(solver="arpack")` calls `eigsh` to get eigenvectors of `A @ A.T.conj()`. However ARPACK has no restarted Lanczos algorithm for complex hermitian matrices, therefore in the case of complex matrices, `eigsh` calls `eigs` and discards the imaginary part of the obtained eigenvectors. This is not satisfying but has been the implementation from the beginning. It is probably less stable and less efficient than a native implementation, but in the easy case of matrix of ones it should still work. I do not understand why this is not the case.

**Comment by lobpcg** [comment]
> 

@ogauthe thanks - I did not know this. I am unsure about the motivation to discards the imaginary part of the eigenvectors of a complex-valued matrix even Hermitian, since the real parts of the complex eigenvectors are no longer eigenvectors? Anyway, it is unrelated to this PR, so we should suppress the error for now and create a new issue to deal with it later?

**Comment by lobpcg** [comment]
The failed tests appear unrelated to this PR. Ready to merge?

**Comment by lobpcg** [comment]
Is there anything else I need to do, other than being patient?

**Comment by tupui** [comment]
> Is there anything else I need to do, other than being patient?

@lobpcg thank you for the updates. As I said above, we are in a release process and this PR not being part of the release, there is no priority and it's totally up to maintainers with the required skills to approve it (my approval would not really have any weight here).

In general 3 days is nothing for PR to big open source projects. Also cf my message above.

**Comment by lobpcg** [comment]
@ilayn - could you please review? The changes are rather trivial and affect only a few lines of the code.

File: scipy/sparse/linalg/_eigen/tests/test_svds.py
Comment (ilayn) [inline scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: We can't do this. If it doesn't work then we have to fix this issue first. We don't have the luxury to push the problems to the users. It doesn't matter if it is related to this PR or not. If there is an issue with ARPACK then this PR is blocked by that problem and until it is resolved we can't proceed with others. There are many PRs waiting in that state.

Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: @ilayn - thank you for your prompt review!
These two lines just basically copy/paste the already skipped complex ARPACK tests in lines 568-569 as well as in lines 743-745. So this is not a new issue - complex ARPACK tests already failed before, - but it has not blocked merging unrelated PRs in the past. 
There are already apparently relevant issues open, e.g., https://github.com/scipy/scipy/issues/13692, https://github.com/scipy/scipy/issues/3716
that hopefully would be addressed at some point, but they do not prevent from making unrelated changes in svds,

Comment (ilayn) [reply scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: Lately, we have been ignoring almost every issue regarding the stability of these in the name of "let's do it in a later PR" and apologies for being so blunt but this module is going down in terms of quality with that kind of behavior. It's collecting lots and lots of edge cases where things are broken and waiting to be picked up again that never happens. But following them up are very difficult because it is full of ignored tests and other comments that are dispersed in many PRs. 

If ARPACK is not working properly or showing its age, then deprecate it. If it is working but we can't make it work then let's fix that part but leaving some solvers in broken state is not going to help anyone. 

I am having difficulty already finding time for contributing to SciPy and doing passing-by reviews and I really don't want to continuously argue with sparse contributors about this. I'll leave it to other maintainers to decide.

Comment (tupui) [reply scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: I second @ilayn's comment. If we have better/more robust solvers, then let's remove this one.

Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: This is not an ARPACK issue. svds(solver="arpack") calls eigsh to get eigenvectors of A @ A.T.conj(). However ARPACK has no restarted Lanczos algorithm for complex hermitian matrices, therefore in the case of complex matrices, eigsh calls something else. 

At any rate, it's a known issue (judging by already skipped unit tests) and clearly unrelated to this PR.

Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/tests/test_svds.py:721]: @tupui @ilayn This discussion is unrelated to the given PR. I propose that the discussion is moved to https://github.com/scipy/scipy/issues/13692#issuecomment-1159782445 where the issue has been reported. I've just made a corresponding comment there.

I also propose that this PR is merged, so that we could move on to fixing the next issue with the script, e.g., https://github.com/scipy/scipy/issues/16330

**Comment by lobpcg** [comment]
Can we please finish this PR somehow? I have already difficulty remembering what it does...

**Comment by lobpcg** [comment]
scipy.scipy (Main Windows Python38-64bit-full) Failing just after merging main, unrelated to this PR

**Comment by lobpcg** [comment]
This trivial PR is finished, has no changes, and is idle since May 31, ready to merge IMHO. Anything else I can do to help it progress?

**Comment by lobpcg** [comment]
Last successful check https://github.com/scipy/scipy/pull/16323/commits/e103f6cc4bc818b569e67b156dd25f72cacedb42 followed by a few [Merge branch 'scipy:main' into patch-1] all failed for some reasons, but clearly unrelated to this PR.

**Comment by j-bowhay** [comment]
The docstring example seems to be failing here:
```
scipy.sparse.linalg.svds
------------------------

File "build/testenv/lib/python3.9/site-packages/scipy/sparse/linalg/_eigen/_svds.py", line 239, in svds
Failed example:
    (np.allclose(s3, s) and
     np.allclose(np.abs(u3), np.abs(u.toarray())) and
     np.allclose(np.abs(vT3), np.abs(vT.toarray())))
Expected:
    True
Got:
    False

```

**Comment by lobpcg** [comment]
> The docstring example seems to be failing here:
> 
> ```
> scipy.sparse.linalg.svds
> ------------------------
> 
> File "build/testenv/lib/python3.9/site-packages/scipy/sparse/linalg/_eigen/_svds.py", line 239, in svds
> Failed example:
>     (np.allclose(s3, s) and
>      np.allclose(np.abs(u3), np.abs(u.toarray())) and
>      np.allclose(np.abs(vT3), np.abs(vT.toarray())))
> Expected:
>     True
> Got:
>     False
> ```

Yes, thanks! I have noticed, too. Strange, since it has always passed before - let me double-check - and no changes are made on my side like for a month, only merging the main...

**Comment by j-bowhay** [comment]
@lobpcg pulled this down locally and the refguide check pasted fine on my machine. Difficult to debug when you cant reproduce the issue! @tupui any suggestions?

**Comment by lobpcg** [comment]
Trying to do it over in #16690...

**Comment by lobpcg** [comment]
Transferred to https://github.com/scipy/scipy/pull/16690

**Pull request closed without merging**
